File: android-wallet/app/src/main/java/com/offlinepay/wallet/Voucher.kt
package com.offlinepay.wallet

import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName
import kotlinx.serialization.json.Json

/// Compact card payload that crosses NFC / QR. v2 schema includes `recipient`
/// so a relay broadcaster can settle on the offline pair's behalf without
/// being able to redirect the funds.
@Serializable
data class CardVoucherPayload(
    @SerialName("v") val version: Int = 2,
    @SerialName("i") val voucherId: String,
    @SerialName("p") val payer: String,
    /// Defaults to the zero address. The encoder is configured with
    /// `encodeDefaults = false`, so when these fields equal the default
    /// they are omitted from the wire JSON. This shrinks bearer cards
    /// (where both are 0x0) by ~100 bytes — the difference between
    /// fitting a MIFARE Classic 1K (336B) or not. Recipient-bound
    /// vouchers set non-zero recipient explicitly, which gets included.
    @SerialName("m") val merchant: String = "0x0000000000000000000000000000000000000000",
    @SerialName("r") val recipient: String = "0x0000000000000000000000000000000000000000",
    @SerialName("a") val amount: String,
    @SerialName("e") val expiry: Long,
    @SerialName("n") val nonce: Long,
    @SerialName("s") val signature: String,
    /// Hardware UID of the MIFARE card / keyfob this voucher was bound to
    /// at top-up. Optional (null for phone-tap flows). Merchant's ESP32
    /// validates `cardUid == card.PICC_ReadCardSerial` before forwarding,
    /// so a voucher copied to a different card won't pass the reader.
    @SerialName("u") val cardUid: String? = null,
)

/// In-memory voucher fields used by verifier + Room DAO. `recipient` is
/// part of the digest now — the address the sender chose at sign time.
data class Voucher(
    val voucherId: String,
    val payer: String,
    val merchant: String,
    val recipient: String,
    val amount: java.math.BigInteger,
    val expiry: Long,
    val nonce: Long,
    val signature: String,
    /// MIFARE/keyfob hardware UID this voucher was bound to. Null for
    /// phone-tap (HCE) and QR/NFC flows. Set only when the voucher was
    /// loaded onto a physical card via `signBearerForCard`.
    val cardUid: String? = null,
) {
    val isTrueBearer: Boolean
        get() = recipient == "0x0000000000000000000000000000000000000000"

    companion object {
        fun fromCardPayload(p: CardVoucherPayload) = Voucher(
            voucherId = p.voucherId,
            payer     = p.payer,
            merchant  = p.merchant,
            recipient = p.recipient,
            amount    = java.math.BigInteger(p.amount),
            expiry    = p.expiry,
            nonce     = p.nonce,
            signature = p.signature,
            cardUid   = p.cardUid,
        )
    }
}

// Important: encodeDefaults = false. CardVoucherPayload has defaults for
// merchant + recipient (both 0x0) so they're omitted from bearer-card
// JSON, shrinking the payload to fit a MIFARE Classic 1K's 336-byte
// data area. Non-bearer (recipient-bound) vouchers carry non-default
// values which get serialized normally. The deserializer round-trips
// correctly because the missing fields fall back to the defaults above.
private val cardJson = Json { encodeDefaults = false; ignoreUnknownKeys = true }

/// Serialize a freshly-signed voucher into the wire JSON. Matches
/// backend/voucher.js:voucherToCardJson byte-for-byte. `cardUid` is
/// included only for bearer cards bound to a specific MIFARE.
fun VoucherSigner.SignedVoucher.toCardJson(cardUid: String? = null): String =
    cardJson.encodeToString(
        CardVoucherPayload.serializer(),
        CardVoucherPayload(
            voucherId = voucherId,
            payer     = payer,
            merchant  = merchant,
            recipient = recipient,
            amount    = amount.toString(),
            expiry    = expiry,
            nonce     = nonce,
            signature = signature,
            cardUid   = cardUid,
        )
    )

fun Voucher.Companion.fromCardJson(json: String): Voucher {
    val payload = cardJson.decodeFromString(CardVoucherPayload.serializer(), json)
    return fromCardPayload(payload)
}

/// Wire payload over NFC is a JSON ARRAY of card payloads (one or more
/// pre-signed bearer vouchers). Tolerates the legacy single-object form too.
fun Voucher.Companion.listFromWireJson(json: String): List<Voucher> {
    val trimmed = json.trim()
    return if (trimmed.startsWith("[")) {
        val arr = cardJson.decodeFromString(
            kotlinx.serialization.builtins.ListSerializer(CardVoucherPayload.serializer()),
            trimmed
        )
        arr.map { fromCardPayload(it) }
    } else {
        listOf(fromCardJson(trimmed))
    }
}

/// Build the JSON-array wire payload from a list of pre-signed cardPayload
/// strings (the exact JSON the backend issued). Avoids re-serialization to
/// preserve the digest-verifying byte order.
fun cardPayloadsToWireJson(cardPayloads: List<String>): String =
    cardPayloads.joinToString(prefix = "[", postfix = "]")

File: backend/src/voucher.js
import { ethers } from "ethers";
import { v4 as uuidv4 } from "uuid";
import { config } from "./config.js";

/// Build the canonical OfflineVault voucher digest.
/// MUST match OfflineVault.voucherDigest exactly:
///   (payer, merchant, recipient, amount, expiry, nonce, voucherId, chainId, vault)
/// `recipient` is signed so relay broadcasters cannot redirect funds.
export function voucherDigest(v) {
  const encoded = ethers.AbiCoder.defaultAbiCoder().encode(
    ["address","address","address","uint256","uint256","uint256","bytes32","uint256","address"],
    [v.payer, v.merchant, v.recipient, v.amount, v.expiry, v.nonce, v.voucherId, config.chainId, config.vault]
  );
  return ethers.keccak256(encoded);
}

/// Sign a voucher with the given ethers Wallet (EIP-191 personal_sign over the digest).
export async function signVoucher(wallet, voucher) {
  const digest = voucherDigest(voucher);
  const sig = await wallet.signMessage(ethers.getBytes(digest));
  return sig;
}

/// Recover the signer address from a signed voucher — used to verify before
/// storing in DB or before forwarding to the merchant.
export function recoverVoucherSigner(voucher, signature) {
  const digest = voucherDigest(voucher);
  return ethers.verifyMessage(ethers.getBytes(digest), signature);
}

/// Generate a fresh voucherId as bytes32.
export function newVoucherId() {
  return ethers.id(uuidv4());
}

/// Build a voucher object (without signature). amountUsdc is in base units (6 decimals).
/// `recipient` is required for bearer vouchers (where merchant=0); the contract
/// rejects recipient=0 on the bearer settle path. For legacy fixed-merchant
/// flows pass recipient = merchant (or 0 — they're not used there).
export function buildVoucher({ payer, merchant, recipient, amountUsdc, ttlSeconds, nonce }) {
  return {
    payer,
    merchant: merchant || ethers.ZeroAddress,
    recipient: recipient || ethers.ZeroAddress,
    amount: BigInt(amountUsdc),
    expiry: BigInt(Math.floor(Date.now() / 1000) + (ttlSeconds || 24 * 3600)),
    nonce: BigInt(nonce),
    voucherId: newVoucherId()
  };
}

/// Compact JSON for the NFC wire / QR code. Matches Voucher.kt CardVoucherPayload.
export function voucherToCardJson(v, signature) {
  return JSON.stringify({
    v: 2,                      // bumped: schema includes recipient
    i: v.voucherId,
    p: v.payer,
    m: v.merchant,
    r: v.recipient,
    a: v.amount.toString(),
    e: Number(v.expiry),
    n: Number(v.nonce),
    s: signature
  });
}

export function voucherFromCardJson(json) {
  const o = JSON.parse(json);
  return {
    voucher: {
      payer: o.p,
      merchant: o.m,
      recipient: o.r,
      amount: BigInt(o.a),
      expiry: BigInt(o.e),
      nonce: BigInt(o.n),
      voucherId: o.i
    },
    signature: o.s
  };
}
